test/baddtlstest.c

```diff
@@ -0,0 +1,603 @@
+        0x00, 0x01, /* Epoch */
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, /* Seq# */
+        0x00, 0x40, /* Length */
+    };
+    static unsigned char finished_msg[] = {
+        0x14, /* Finished */
+        0x00, 0x00, 0x0c, /* Length */
+        0x00, 0x03, /* Seq# */
+        0x00, 0x00, 0x00, /* Fragment offset */
+        0x00, 0x00, 0x0c, /* Fragment length */
+        /* Finished MAC (12 bytes) */
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00,
+        /* Record MAC (20 bytes) */
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+        0x00, 0x00, 0x00, 0x00,
+        /* Padding to 48 bytes */
+        0x03, 0x03, 0x03, 0x03
+    };
+    unsigned char record_iv[16];
+    unsigned char handshake_hash[EVP_MAX_MD_SIZE];
+
+    /* Derive key material */
+    do_PRF(TLS_MD_KEY_EXPANSION_CONST, TLS_MD_KEY_EXPANSION_CONST_SIZE,
+           server_random, SSL3_RANDOM_SIZE,
+           client_random, SSL3_RANDOM_SIZE,
+           key_block, sizeof(key_block));
+
+    /* Generate Finished MAC */
+    if (!EVP_DigestFinal_ex(handshake_md, handshake_hash, NULL))
+        printf("EVP_DigestFinal_ex() failed\n");
+
+    do_PRF(TLS_MD_SERVER_FINISH_CONST, TLS_MD_SERVER_FINISH_CONST_SIZE,
+           handshake_hash, EVP_MD_CTX_size(handshake_md),
+           NULL, 0,
+           finished_msg + 12, 12);
+
+    /* We want to use ssl3_cbc_digest_record() here but it's not
+     * available if we're testing against a shared library incarnation
+     * of OpenSSL. So we do it ourselves. */
+    hmac_packet(finished_hdr, finished_msg, 24);
+    /* We did the padding manually to a multiple of 16 bytes */
+    encrypt_packet(record_iv, finished_msg, sizeof(finished_msg));
+
+    BIO_write(rbio, finished_hdr, sizeof(finished_hdr));
+    BIO_write(rbio, record_iv, sizeof(record_iv));
+    BIO_write(rbio, finished_msg, sizeof(finished_msg));
+
+    return 1;
+}
+
+static int validate_ccs(BIO *wbio)
+{
+    long len;
+    unsigned char *data;
+    unsigned char *dataend;
+
+    len = BIO_get_mem_data(wbio, (char **)&data);
+    dataend = data + len;
+
+    /* Skip the record header */
+    data += DTLS1_RT_HEADER_LENGTH;
+        /* Check client version */
+    if (data + 2 > dataend)
+        return 0;
+    if (data[0] != 0x01 || data[1] != 0x00)
+        return 0;
+
+    if (data + 3 > dataend)
+        return 0;
+    /* Check CCS message content, including seq# 2 */
+    if (data[0] != 0x01 || data[1] != 0x00 || data[2] != 0x02)
+        return 0;
+    data += 3; /* DTLS1_BAD_VER CCS length */
+
+    /* We expect the Finished message here.*/
+    if (data + DTLS1_RT_HEADER_LENGTH > dataend)
+        return 0;
+    /* Check record header type, version */
+    if (data[0] != 0x16 || data[1] != 0x01 || data[2] != 0x00)
+        return 0;
+    /* Check epoch is now 1 */
+    if (data[3] != 00 || data[4] != 0x01)
+        return 0;
```

Comment:
Yep. If you accepted my Finished then you are evidently getting the MAC right so far. Implementing decryption just to check you get it right for one more packet seems pointless. I don't need to check that you know how to craft a proper Finished packet; that is not a DTLS1_BAD_VER special thing so is tested elsewhere.